gen/emit/EmitJulia.cpp

```diff
@@ -70,47 +70,91 @@ struct EmitJulia : public Emitter
         }
     }
 
+    string topLevelPackage(const string& package) {
+        auto packageParts = StringUtil::split(package, '.');
+        if (packageParts.size() > 0) {
+            return packageParts.at(0);
+        } else {
+            return "";
+        }
+    }
+
     void emitDependencies()
     {
-        unordered_set<string> dependencies;
+        unordered_set<string> imports;
+        string lsPackage = topLevelPackage(ls.structname.package);
+        if (lsPackage != "") {
+            imports.insert("import " + lsPackage);
+        }
+
         for (auto& lm : ls.members) {
             auto& tn = lm.type.fullname;
             if (!ZCMGen::isPrimitiveType(tn) &&
-                dependencies.find(tn) == dependencies.end() &&
+                imports.find(tn) == imports.end() &&
                 tn != ls.structname.fullname) {
-                dependencies.insert(lm.type.nameUnderscore());
+                string lmPackage = topLevelPackage(lm.type.package);
+                if (lmPackage != "") {
+                    // The dependent type belongs to some package
+                    if (lmPackage == lsPackage) {
+                        // Then this is a dependency on a package from somewhere
+                        // in the same module, so we should import its outer
+                        // Julia module
+                        imports.insert("import _" + lm.type.nameUnderscore());
+                    } else {
+                        // Then this is a dependency on some other package, so
+                        // we can just import the entire other package:
+                        imports.insert("import " + lmPackage);
+                    }
+                } else {
+                    // The dependent type does not belong to any package, so
+                    // we just import it from its outer module
+                    imports.insert("import _" + lm.type.shortname + ": " + lm.type.shortname);
+                }
             }
         }
 
-        for (auto& tn : dependencies) {
-            emit(0, "include(\"%s.jl\")", tn.c_str());
+        if (lsPackage != "") {
+            // The type we're creating is in a package, so we need to @eval its
+            // definition into the right module
+            emit(0, "import %s", lsPackage.c_str());
+            emit(0, "@eval %s begin", ls.structname.package.c_str());
+        } else {
+            // The type we're creating is not in a package, so we don't need to
+            // do anything. But we'll create a "begin" block so that the number
+            // of terminating "end" statements is the same no matter which path
+            // was chosen here.
+            emit(0, "begin");
```

Comment:
Should we be exporting the type from the module? I'm familiar enough with Julia package development to know if that's the more language appropriate way to expose the type to the user.

Reply:
yes, I think exporting the type is totally reasonable. I'll do that